### PR TITLE
Offline driver initialisation refactor continued (cable_user%MetType)

### PR DESCRIPTION
**Pull request by SeanBryan51**
This change further consolidates the initialisation logic within the offline drivers (in particular logic related to `cable_user%MetType`) to remove code duplication and improve code clarity across each of the drivers (see #425, #440, #469 and #472).

As part of the refactoring changes, the `cable_user%MetType = 'gpgs'` namelist parameter is deprecated as it is equivalent to setting `cable_user%MetType = 'gswp'` and `leaps = .TRUE.`.

## Type of change

- [x] Code refactor
- [x] New or updated documentation

## Checklist

- [x] The new content is accessible and located in the appropriate section.
- [x] I have checked that links are valid and point to the intended content.
- [x] I have checked my code/text and corrected any misspellings

----
📚 Documentation preview 📚: https://cable--498.org.readthedocs.build/en/498/

### CODE REVIEW

**Comment by SeanBryan51** [comment]
I won't have time to move the remaining occurrences of `CALL1` to a shared initialisation routine before going on leave next week but I have noted them below:

1. https://github.com/CABLE-LSM/CABLE/blob/7e6a8bb83321b764bb333fd02bbbf4b4cbf2b9ee/src/offline/cable_serial.F90#L367-L444
2. https://github.com/CABLE-LSM/CABLE/blob/7e6a8bb83321b764bb333fd02bbbf4b4cbf2b9ee/src/offline/cable_serial.F90#L483-L487
3. https://github.com/CABLE-LSM/CABLE/blob/7e6a8bb83321b764bb333fd02bbbf4b4cbf2b9ee/src/offline/cable_serial.F90#L490-L494
4. https://github.com/CABLE-LSM/CABLE/blob/7e6a8bb83321b764bb333fd02bbbf4b4cbf2b9ee/src/offline/cable_mpimaster.F90#L384-L420
5. https://github.com/CABLE-LSM/CABLE/blob/7e6a8bb83321b764bb333fd02bbbf4b4cbf2b9ee/src/offline/cable_mpimaster.F90#L424-L526
6. https://github.com/CABLE-LSM/CABLE/blob/7e6a8bb83321b764bb333fd02bbbf4b4cbf2b9ee/src/offline/cable_mpiworker.F90#L278-L285
7. https://github.com/CABLE-LSM/CABLE/blob/7e6a8bb83321b764bb333fd02bbbf4b4cbf2b9ee/src/offline/cable_mpiworker.F90#L291-L419

File: src/offline/cable_driver_init.F90 [hunk 1]
```diff
@@ -189,4 +208,124 @@ SUBROUTINE cable_driver_init(mpi_grp, trunk_sumbal, NRRRR)
...
+        WRITE(logn,*) 'undefined start year for gswp met: '
+        WRITE(logn,*) 'enter value for ncciy or'
+        WRITE(logn,*) '(CABLE_USER%YearStart and  CABLE_USER%YearEnd) in cable.nml'
+        STOP
+      END IF
+      cable_user%YearStart = ncciy
+      cable_user%YearEnd = ncciy
+    END IF
+
+    IF (.NOT. (PRESENT(GSWP_MID) .AND. PRESENT(NRRRR))) return
```
Comment (SeanBryan51) [inline src/offline/cable_driver_init.F90:233]:
```suggestion
    IF (.NOT. (PRESENT(GSWP_MID) .AND. PRESENT(NRRRR))) RETURN
```

**Comment by ccarouge** [comment]
@SeanBryan51 I know you are having issues testing all the changes but have you tried running benchab? The sites and the spatial runs should return identical results before and after.

**Comment by ccarouge** [comment]
For `gpgs`, please add it as deprecated on this page: https://cable.readthedocs.io/en/latest/user_guide/other_resources/obsolete_and_deprecated_features/obsolete_and_deprecated_features/ In replacement, just put the "gswp + leap" reason.

**Comment by SeanBryan51** [comment]
Running it through benchcab now.

**Comment by SeanBryan51** [comment]
It looks like I forgot to update the stubs for the serial build - but it also made me realise the build-ci currently builds only the MPI executable (due to [the default value of the `mpi` variant](https://github.com/ACCESS-NRI/spack-packages/blob/512b0e0940c798c712f6535624c537d5c55ccb56/packages/cable/package.py#L29) in the CABLE spack package definition). Something to note for the future.

**Review by ccarouge — COMMENTED** [review]
Seems ok but let's see what benchcab thinks of it.

**Comment by SeanBryan51** [comment]
Benchcab runs are bitwise reproducible with `main`:

PBS log file:

```
2024-11-28 15:40:15,127 - INFO - benchcab.benchcab.py:380 - Running comparison tasks...
2024-11-28 15:40:15,151 - INFO - benchcab.benchcab.py:381 - tasks: 168 (models: 2, sites: 42, science configurations: 4)
2024-11-28 15:43:03,606 - INFO - benchcab.benchcab.py:391 - 0 failed, 168 passed
```

Benchcab version: `4.2.0`

`config.yaml`:
```yaml
realisations:
  - repo:
      git:
        branch: main
  - repo:
      git:
        branch: 425-offline-driver-initialisation-refactor-part3

modules: [
  intel-compiler/2021.1.1,
  netcdf/4.7.4,
  openmpi/4.1.0
]
```

Spatial (MPI) runs all execute as expected (note outputs do not reproduce as per issue #463).

**Review by ccarouge — APPROVED** [review]
It's good for me.

File: src/offline/cable_mpimaster.F90 [hunk 2]
```diff
@@ -391,48 +342,8 @@ SUBROUTINE mpidrv_master (comm, trunk_sumbal)
...
-                timeunits="seconds since "//TRIM(str1)//"-"//TRIM(str2)//"-"//TRIM(str3)//" &
-                     00:00"
-
-             ENDIF
              kend = NINT(24.0*3600.0/dels) * LOY
           ELSE IF ( TRIM(cable_user%MetType) .EQ. 'cru' ) THEN
-             ! CLN HERE CRU modfications
-             IF ( CALL1 ) THEN
-
-                CALL CPU_TIME(etime)
```
Comment (micaeljtoliveira) [inline src/offline/cable_mpimaster.F90:415]: I couldn't see the corresponding calls to `CPU_TIME` in the new code. Did they disappear or is it github confusing me?

**Review by micaeljtoliveira — COMMENTED** [review]
@SeanBryan51 I had another look and I have one suggestion and one question. Otherwise looks fine!

File: src/offline/cable_mpimaster.F90 (hunk 2 above)
Comment (SeanBryan51) [reply src/offline/cable_mpimaster.F90:415]: The calls to `CPU_TIME(etime)` were removed from the initialisation stages as the value of `etime` was not used in any of the drivers - `etime` was only used after a subsequent call to `CPU_TIME` which overwrites the previous value of `etime` from the initialisation stage so they were safe to remove.

**Review by micaeljtoliveira — APPROVED** [review]

File: src/offline/cable_driver_init.F90 [hunk 3]
```diff
@@ -147,6 +162,10 @@ SUBROUTINE cable_driver_init(mpi_grp, trunk_sumbal, NRRRR)
       CALL set_group_output_values()
     END IF
 
+    IF (TRIM(cable_user%POPLUC_RunType) == 'static') THEN
+      cable_user%POPLUC= .FALSE.
+    END IF
```
[the remark this answers is not in the document]
Comment (SeanBryan51) [reply src/offline/cable_driver_init.F90:167]: The truth table for your suggestion slightly differs to the original code when `cable_user%POPLUC` is `.FALSE.` and `cable_user%POPLUC_RunType == 'static'` is `.FALSE.`. In this case, your suggestion will set `cable_user%POPLUC = .TRUE.` where as the original code will leave `cable_user%POPLUC` set to `.FALSE`. The original code would be equivalent to:
```fortran
cable_user%POPLUC = cable_user%POPLUC .AND. TRIM(cable_user%POPLUC_RunType) /= 'static'
```
I think I will leave the code as is as the behaviour looks clearer than the code above.

[the remark this answers is not in the document]
Comment (SeanBryan51) [reply src/offline/cable_driver_init.F90:167]: I suppose the real question here is why `cable_user%POPLUC_RunType = 'static'` is even an option to begin with. I'm not sure why this is the case.

[the remark this answers is not in the document]
Comment (micaeljtoliveira) [reply src/offline/cable_driver_init.F90:167]: You are absolutely right. Sorry, I should have read the code more carefully.

**Pull request merged**